docs/madr/decisions/046-policy-on-namespace.md

```diff
@@ -0,0 +1,456 @@
+                name: backend
+                tags:
+                  version: v2
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: backend-route-timeout
+  namespace: kuma-system
+spec:
+  targetRef:
+    kind: MeshHTTPRoute
+    name: backend-route
+  to:
+    - targetRef:
+        kind: Mesh
+      default:
+        http:
+          requestTimeout: 5s
+```
+
+**2. MeshTimeout in "kuma-system", MeshHTTPRoute in "frontend-ns"**
+
+Frontend Service Owner creates a MeshHTTPRoute in the `frontend-ns` namespace.
+What are the reasons for Mesh Operator to create MeshTimeout in `kuma-system` and reference the MeshHTTPRoute policy
+in `frontend-ns`?
+Mesh Operator should work with policies that apply globally, knowing specifics of the `frontend` app seems incorrect.
+We didn't find good use cases why this type of reference would make sense.
+
+**3. MeshTimeout in "frontend-ns", MeshHTTPRoute in "kuma-system"**
+
+**As** a Backend Service Owner<br>
+**I want** to route all requests with `/v2` prefix to `version: v2` instances<br>
+**so that** requests that need v2 version of the API ended up on correct instances.
+
+**As** a Backend Service Owner<br>
+**I want** all requests routed to `version: v2` instances to have a 5s timeout<br>
+**so that** consumers didn't wait longer than max possible processing time.
+
+**As** a Frontend Service Owner<br>
+**I want** all requests routed to `version: v2` instances of `backend` to have 3s timeout<br>
+**so that** my service meets SLO and responds withing 3s.
+
+This can be achieved by applying 3 policies:
+
+```yaml
+# on kuma-system namespace
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: backend-route
+  namespace: kuma-system
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+        kind: MeshService
+        name: backend
+      rules:
+        - matches:
+            - path:
+                type: PathPrefix
+                value: /v2
+          default:
+            backendRefs:
+              - kind: MeshServiceSubset
+                name: backend
+                tags:
+                  version: v2
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: backend-route-timeout
+  namespace: kuma-system
+spec:
+  targetRef:
+    kind: MeshHTTPRoute
+    name: backend-route
+  to:
+    - targetRef:
+        kind: Mesh
+      default:
+        http:
+          requestTimeout: 5s
+# on frontend-ns
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: frontend-to-backend
+  namespace: frontend-ns
+spec:
+  targetRef:
+    kind: MeshHTTPRoute
+    name: backend-route
+    namespace: kuma-system
+  to:
+    - targetRef:
+        kind: Mesh
+      default:
+        http:
+          requestTimeout: 3s
+```
+
+This example also demonstrates the scenario when 2 MeshTimeout reference the same route,
+the namespaced MeshTimeout is considered more specific.
+That's why the final request timeout value should be 3s.
```

Comment:
That's the biggest question of this MADR.

It is a bit strange that the service owner can advertise their service using their own namespace, but cannot advertise routes to reference in the policies

Reply:
Good point!

So if Backend SO can create `MeshService` in `backend-ns` and advertise it to all other namespaces, should this be the case for MeshHTTPRoutes in `backend-ns`?

Reasons why it makes sense:
* we can think of routes as "a more granular destination than MeshService"
* we already want to do this for namespace-scoped MeshHTTPRoutes that target MeshGateaway
* better mapping to GAMMA producer and consumer routes

Reasons why it doesn't make sense:
* all other namespace-scoped policies affects only pods in the namespace they're applied, it won't be the case for MeshHTTPRoute (or MeshTCPRoute)

@Automaat @michaelbeaumont I'm curious to hear your opinion on this

Reply:
> So if Backend SO can create MeshService in backend-ns and advertise it to all other namespaces, should this be the case for MeshHTTPRoutes in backend-ns?

I think so. I see `kuma-system` as a sort of crutch because we didn't have namespaced policies figured out. Ultimately it shouldn't "need" to be used.

> all other namespace-scoped policies affects only pods in the namespace they're applied

I don't think this is an especially meaningful idea. Like Jakub said. Yes, the namespace-scoped policies affect Pods of other namespaces, but doesn't even the creation of Pods in a given namespace affect Pods of other namespaces?

Reply:
Put differently, what the namespapce-scoped policy affects is _requests_ that would be routed to the services of that namespace. I think this is a more useful abstraction than "does this affect the Envoy config of my Pod?"

Reply:
> I don't think this is an especially meaningful idea

I think it is. For me it's easier to think about policies and impact with this model in mind.

Maybe the problem here is that MeshHTTPRoute is not really a policy, just like a Service is not a policy.

Reply:
But how do you define only "affecting Pods" in a namespace exactly? In light of the fact that I can simply create a Service in another namespace in order to affect them.

I mean at this point we're basically just playing a semantic game where we define a policy as a thing that only affects Envoy config of Pods in its namespace. And a bunch of other things affect Envoy config of other namespaces and those are "not policies". Starting to feel like the term isn't especially useful.

Reply:
It's just we've been here with `sources & destinations`. It was not clear how to affect what and where. Maybe you are right, maybe we need to ditch "all other namespace-scoped policies affects only pods in the namespace they're applied" rule, but I'd be curious to hear what's the alternative so that it is still sane to understand what is matching where.

> I propose that a service owner should never need the mesh operator/access to kuma-system to influence requests to its Services.

That's a _recommendation use-case_.
That's reasonable as long as clients still have a way of overriding this.

Reply:
> That's reasonable as long as clients still have a way of overriding this.

Definitely, the other piece would be if I create a policy in a namespace that targets Services of _different_ namespaces, it affects only requests from the policy's namespace.

Reply:
After reading all that comments and revisiting GAMMA docs, I understand why it would be nice to not use system policy at all. 

So the precedence of applying policies will look like: `kuma-system -> different-ns -> my-ns`

I think that there are two different use cases, first with routes and targeting routes, and the second is rest of the policies.

## First use case (routes):
Example producer route with default timeout, created in `backend-ns`:
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: backend-route
  namespace: backend-ns
spec:
  targetRef:
    kind: MeshService
    name: frontend
  to:
    - targetRef:
        kind: MeshService
        name: backend
        namespace: backend-ns
--- 
apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: frontend-to-backend
  namespace: backend-ns
spec:
  targetRef:
    kind: MeshHTTPRoute
    name: backend-route
    namespace: backend-ns 
  to:
    - targetRef:
        kind: Mesh
      default:
        http:
          requestTimeout: 5s
```

Then frontend SO can override this timeout using:
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: frontend-to-backend
  namespace: frontend-ns # I would like to override default set by producer
spec:
  targetRef:
    kind: MeshHTTPRoute
    name: backend-route
    namespace: backend-ns 
  to:
    - targetRef:
        kind: Mesh
      default:
        http:
          requestTimeout: 2s
```

## Policies not targeting routes

As a backend SO, I would like to create default MeshRetry for requests to my service:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshRetry
metadata:
  name: all-to-backend-retry-http
  namespace: backend-ns
spec:
  targetRef:
    kind: Mesh
    name: default
  to:
    - targetRef:
        kind: MeshService
        name: backend
        namespace: backend-ns
      default:
        http:
          numRetries: 5
``` 
and then frontend SO would like to override this default:
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshRetry
metadata:
  name: to-backend-retry-http
  namespace: frontend-ns
spec:
  targetRef:
    kind: Mesh
    name: default
  to:
    - targetRef:
        kind: MeshService
        name: backend
        namespace: backend-ns
      default:
        http:
          numRetries: 2
```
Should this work? Should first policy be applied to dpps in different namespaces? Or should backend SO create it in `kuma-system` namespace?

@michaelbeaumont Does this make sense? And is it as you would expect it to be?

Reply:
Yeah it does.

> kuma-system -> different-ns -> my-ns

To be clear I was imagining `different-ns` overriding `my-ns` overriding `kuma-system`

Reply:
Why should `different-ns` override `my-ns`? Shouldn't my config, that I just created in my namespace, be most important? 

And there is question about second use case, should it work? 

Also, @jakubdyszkiewicz it would be nice to hear your opinion

Reply:
> Why should different-ns override my-ns? Shouldn't my config, that I just created in my namespace, be most important?

I think it depends on whether routes are meant to be defaults or authoritative.

Reply:
@Automaat when you say:

> So the precedence of applying policies will look like: kuma-system -> different-ns -> my-ns

are you saying "my" from the point of view of producer (Backend SO) or consumer (Frontend SO)?

Reply:
Exactly as Mike explained

Reply:
So if we say that `my-ns` is service owner's ns and `different-ns` is basically consumer's ns, then I agree with Mike – `different-ns` should override `my-ns`. This relative naming is a bit confusing but I feel like we're all on the same page here :)

Reply:
There was a mix up in the naming, but in the end I think we all agree, I was thinking from a perspective of consumer, owner of `my-ns` so `different-ns` was namespace where service I want to consume, route and default timeout was configured.

Reply:
I updated MADR as we discussed